packages/sit-onyx/src/components/OnyxSwitch/OnyxSwitch.vue

```diff
@@ -0,0 +1,229 @@
+<script lang="ts" setup>
+import { OnyxIcon } from "@/index";
+import checkSmall from "@sit-onyx/icons/check-small.svg?raw";
+import xSmall from "@sit-onyx/icons/x-small.svg?raw";
+import type { OnyxSwitchProps } from "./types";
+import { areObjectsFlatEqual } from "@/utils/comparator";
+import { transformValidityStateToObject } from "@/utils/forms";
+import { computed, ref, toRefs, watch } from "vue";
+const props = withDefaults(defineProps<OnyxSwitchProps>(), {
+  modelValue: false,
+  disabled: false,
+  required: false,
+});
+
+const emit = defineEmits<{
+  /** Emitted when the checked state changes. */
+  "update:modelValue": [value: boolean];
+  /** Emitted whenever the validity state of the input changes */
+  validityChange: [state: ValidityState];
+}>();
+
+const { errorMessage } = toRefs(props);
+const inputElement = ref<HTMLInputElement | null>(null);
+const validityState = ref(inputElement.value?.validity);
+
+const isChecked = computed({
+  get: () => props.modelValue,
+  set: (value) => {
+    emit("update:modelValue", value);
+  },
+});
+
+watch([inputElement, errorMessage], () => {
+  if (!inputElement.value) return;
+  // by using setCustomValidity, the ValidityState will turn invalid
+  // as long as it is not an empty string
+  inputElement.value.setCustomValidity(props.errorMessage || "");
+});
+
+watch(
+  [inputElement, isChecked, errorMessage],
+  () => {
+    if (!inputElement.value) return;
+
+    const newValidityState = transformValidityStateToObject(inputElement.value!.validity);
+    //  only update + emit the validity state when it changed
+    if (!validityState.value || !areObjectsFlatEqual(newValidityState, validityState.value)) {
+      validityState.value = newValidityState;
+      emit("validityChange", validityState.value);
+    }
+  },
+  { immediate: true },
+);
+</script>
+
+<template>
+  <label
+    tabindex="0"
+    class="onyx-switch"
+    :class="{ 'onyx-required-marker': props.required, 'onyx-optional-marker': !props.required }"
+  >
+    <input
+      ref="inputElement"
+      v-model="isChecked"
+      class="onyx-switch__input"
+      type="checkbox"
+      :aria-label="props.hideLabel ? props.label : undefined"
+      :disabled="props.disabled"
+      :required="props.required"
+    />
+    <span class="onyx-switch__container">
+      <span class="onyx-switch__icon">
+        <OnyxIcon :icon="isChecked ? checkSmall : xSmall" size="24px" />
+      </span>
+    </span>
+    <span v-if="!props.hideLabel" class="onyx-switch__label">{{ props.label }}</span>
+  </label>
+</template>
+
+<style lang="scss">
+.onyx-switch {
+  display: inline-flex;
+  align-items: center;
+  cursor: pointer;
+  margin: var(--onyx-spacing-2xs);
+  height: 1.5rem;
+
+  &__input {
+    position: absolute;
```

Comment:
Is absolute needed? Does it even have an effect without defining top/left? Also the parent should be relative then, right?

Reply:
When I remove the position it makes the label wider

Reply:
Ah this is the native checkbox, right :) Maybe we can add a small comment that position is needed to hide the native checkbox?

Reply:
Yup, I'll do that